File: src/tradebot_sci/market/indicators.py
from __future__ import annotations
import statistics
from typing import List, Tuple
from tradebot_sci.market.models import Candle

def calculate_sma(data: List[float], period: int) -> float:
    if len(data) < period:
        return 0.0
    return sum(data[-period:]) / period

def calculate_ema(data: List[float], period: int) -> float:
    """Standard Exponential Moving Average."""
    if not data:
        return 0.0
    if len(data) < period:
        return sum(data) / len(data)
    
    alpha = 2 / (period + 1)
    ema = sum(data[:period]) / period # Start with SMA
    for price in data[period:]:
        ema = (price * alpha) + (ema * (1 - alpha))
    return ema

def calculate_bollinger_bands(data: List[float], period: int = 20, std_dev: float = 2.0) -> Tuple[float, float, float]:
    """Returns (Lower, Middle, Upper) bands."""
    if len(data) < period:
        return 0.0, 0.0, 0.0
    
    window = data[-period:]
    middle = sum(window) / period
    stdev = statistics.stdev(window)
    
    lower = middle - (std_dev * stdev)
    upper = middle + (std_dev * stdev)
    
    return lower, middle, upper

def calculate_rsi(data: List[float], period: int = 14) -> float:
    if len(data) < period + 1:
        return 50.0
        
    gains = []
    losses = []
    
    for i in range(1, len(data)):
        diff = data[i] - data[i-1]
        if diff > 0:
            gains.append(diff)
            losses.append(0)
        else:
            gains.append(0)
            losses.append(abs(diff))
            
    # Wilder's Smoothing
    avg_gain = sum(gains[:period]) / period
    avg_loss = sum(losses[:period]) / period
    
    for i in range(period, len(gains)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period
    
    if avg_loss == 0:
        return 100.0
        
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def calculate_macd(data: List[float], fast: int = 12, slow: int = 26, signal: int = 9) -> Tuple[float, float, float]:
    """
    Calculate MACD indicator.
    Returns (macd_line, signal_line, histogram).
    """
    if len(data) < slow + signal:
        return 0.0, 0.0, 0.0

    # Calculate fast and slow EMAs
    fast_ema = calculate_ema(data, fast)
    slow_ema = calculate_ema(data, slow)
    macd_line = fast_ema - slow_ema

    # Build MACD line series for signal EMA
    macd_series = []
    alpha_fast = 2 / (fast + 1)
    alpha_slow = 2 / (slow + 1)

    ema_f = sum(data[:fast]) / fast
    ema_s = sum(data[:slow]) / slow

    for i in range(slow, len(data)):
        if i >= fast:
            ema_f = (data[i] * alpha_fast) + (ema_f * (1 - alpha_fast))
        ema_s = (data[i] * alpha_slow) + (ema_s * (1 - alpha_slow))
        macd_series.append(ema_f - ema_s)

    if len(macd_series) < signal:
        return macd_line, 0.0, macd_line

    # Signal line = EMA of MACD series
    sig_alpha = 2 / (signal + 1)
    sig_ema = sum(macd_series[:signal]) / signal
    for val in macd_series[signal:]:
        sig_ema = (val * sig_alpha) + (sig_ema * (1 - sig_alpha))

    histogram = macd_line - sig_ema
    return macd_line, sig_ema, histogram


def calculate_macd_series(data: List[float], fast: int = 12, slow: int = 26, signal: int = 9) -> Tuple[List[float], List[float], List[float]]:
    """
    Calculate full MACD series for crossover detection.
    Returns (macd_lines[], signal_lines[], histograms[]).
    """
    if len(data) < slow + signal:
        return [], [], []

    alpha_fast = 2 / (fast + 1)
    alpha_slow = 2 / (slow + 1)
    sig_alpha = 2 / (signal + 1)

    ema_f = sum(data[:fast]) / fast
    ema_s = sum(data[:slow]) / slow

    macd_series = []
    for i in range(slow, len(data)):
        if i >= fast:
            ema_f = (data[i] * alpha_fast) + (ema_f * (1 - alpha_fast))
        ema_s = (data[i] * alpha_slow) + (ema_s * (1 - alpha_slow))
        macd_series.append(ema_f - ema_s)

    if len(macd_series) < signal:
        return macd_series, [], []

    # Build signal series
    sig_ema = sum(macd_series[:signal]) / signal
    signal_series = [sig_ema]
    for val in macd_series[signal:]:
        sig_ema = (val * sig_alpha) + (sig_ema * (1 - sig_alpha))
        signal_series.append(sig_ema)

    # Align: trim macd_series to match signal_series length
    aligned_macd = macd_series[signal - 1:]
    histograms = [m - s for m, s in zip(aligned_macd, signal_series)]

    return aligned_macd, signal_series, histograms


def calculate_vwap(candles: List[Candle]) -> float:
    """
    Calculate Volume-Weighted Average Price from candle data.
    Uses typical price (H+L+C)/3 weighted by volume.
    """
    if not candles:
        return 0.0

    total_vp = 0.0
    total_volume = 0.0

    for c in candles:
        vol = getattr(c, 'volume', 0) or 0
        if vol <= 0:
            continue
        typical_price = (c.high + c.low + c.close) / 3.0
        total_vp += typical_price * vol
        total_volume += vol

    if total_volume == 0:
        # Fallback: simple average of closes
        closes = [c.close for c in candles]
        return sum(closes) / len(closes) if closes else 0.0

    return total_vp / total_volume

File: src/tradebot_sci/electron_gui/indicators.js
// ═══════════════════════════════════════════════════════════════
//  indicators.js — Technical indicator calculations
//  Extracted from renderer.js for maintainability.
// ═══════════════════════════════════════════════════════════════

window.Indicators = (() => {

    function calculateSMA(data, period) {
        const result = [];
        for (let i = period - 1; i < data.length; i++) {
            let sum = 0;
            for (let j = 0; j < period; j++) {
                sum += data[i - j].close;
            }
            result.push({ time: data[i].time, value: sum / period });
        }
        return result;
    }

    function calculateEMA(data, period) {
        const result = [];
        const multiplier = 2 / (period + 1);

        // Start with SMA for the first EMA value
        if (data.length < period) return result;

        let sum = 0;
        for (let i = 0; i < period; i++) {
            sum += data[i].close;
        }
        let ema = sum / period;
        result.push({ time: data[period - 1].time, value: ema });

        for (let i = period; i < data.length; i++) {
            ema = (data[i].close - ema) * multiplier + ema;
            result.push({ time: data[i].time, value: ema });
        }
        return result;
    }

    function calculateBollingerBands(data, period = 20, stdDevMultiplier = 2) {
        const upper = [], middle = [], lower = [];
        for (let i = period - 1; i < data.length; i++) {
            let sum = 0;
            for (let j = 0; j < period; j++) sum += data[i - j].close;
            const mean = sum / period;
            let sqSum = 0;
            for (let j = 0; j < period; j++) sqSum += Math.pow(data[i - j].close - mean, 2);
            const stdDev = Math.sqrt(sqSum / period);
            const t = data[i].time;
            upper.push({ time: t, value: mean + stdDevMultiplier * stdDev });
            middle.push({ time: t, value: mean });
            lower.push({ time: t, value: mean - stdDevMultiplier * stdDev });
        }
        return { upper, middle, lower };
    }

    function calculateVWAP(data) {
        // data must have .close and .volume (use rawCandleData)
        const result = [];
        let cumVol = 0, cumTP = 0;
        for (let i = 0; i < data.length; i++) {
            const tp = (data[i].high + data[i].low + data[i].close) / 3;
            const vol = data[i].volume || 0;
            cumVol += vol;
            cumTP += tp * vol;
            if (cumVol > 0) {
                result.push({ time: data[i].time, value: cumTP / cumVol });
            }
        }
        return result;
    }

    function calculateRSI(data, period = 14) {
        const result = [];
        if (data.length < period + 1) return result;
        let gains = 0, losses = 0;
        for (let i = 1; i <= period; i++) {
            const change = data[i].close - data[i - 1].close;
            if (change >= 0) gains += change;
            else losses -= change;
        }
        let avgGain = gains / period;
        let avgLoss = losses / period;
        const rsi = avgLoss === 0 ? 100 : 100 - (100 / (1 + avgGain / avgLoss));
        result.push({ time: data[period].time, value: rsi });
        for (let i = period + 1; i < data.length; i++) {
            const change = data[i].close - data[i - 1].close;
            avgGain = (avgGain * (period - 1) + (change >= 0 ? change : 0)) / period;
            avgLoss = (avgLoss * (period - 1) + (change < 0 ? -change : 0)) / period;
            const r = avgLoss === 0 ? 100 : 100 - (100 / (1 + avgGain / avgLoss));
            result.push({ time: data[i].time, value: r });
        }
        return result;
    }

    function calculateHeikinAshi(data) {
        if (!data || data.length === 0) return [];
        const result = [];
        for (let i = 0; i < data.length; i++) {
            const c = data[i];
            const haClose = (c.open + c.high + c.low + c.close) / 4;
            const haOpen = i === 0
                ? (c.open + c.close) / 2
                : (result[i - 1].open + result[i - 1].close) / 2;
            const haHigh = Math.max(c.high, haOpen, haClose);
            const haLow = Math.min(c.low, haOpen, haClose);
            result.push({ time: c.time, open: haOpen, high: haHigh, low: haLow, close: haClose });
        }
        return result;
    }

    return {
        calculateSMA,
        calculateEMA,
        calculateBollingerBands,
        calculateVWAP,
        calculateRSI,
        calculateHeikinAshi,
    };
})();
